subiquity/server/controllers/tests/test_filesystem.py

```diff
@@ -457,3 +458,68 @@ async def test_lvm_20G_bad_offset(self, disk_size):
         self.assertEqual(
                 disk_size - (1 << 20), parts[-1].offset + parts[-1].size,
                 disk_size)
+
+
+class TestApplySystem(TestCase):
+
+    def setUp(self):
+        self.app = make_app()
+        self.app.opts.bootloader = 'UEFI'
+        self.app.report_start_event = mock.Mock()
+        self.app.report_finish_event = mock.Mock()
+        self.app.prober = mock.Mock()
+        self.fsc = FilesystemController(app=self.app)
+        self.fsc._configured = True
+        self.fsc.model = make_model(Bootloader.UEFI)
+
+    def test_add_structure(self):
+        disk = make_disk(self.fsc.model)
+        self.fsc._add_structure(
+            disk=disk,
+            offset=100,
+            size=2 << 20,
+            is_last=False,
+            structure=snapdapi.VolumeStructure(
+                name='ptname',
+                type="83,0FC63DAF-8483-4772-8E79-3D69D8477DE4",
+                label='label',
+                size=1 << 20,
+                role=snapdapi.Role.SYSTEM_DATA,
+                filesystem='ext4'))
+        [part] = disk.partitions()
+        self.assertEqual(part.offset, 100)
+        self.assertEqual(part.partition_name, 'ptname')
+        self.assertEqual(part.flag, 'linux')
+        self.assertEqual(part.size, 2 << 20)
+        self.assertEqual(part.fs().fstype, 'ext4')
+        self.assertEqual(part.fs().mount().path, '/')
+        self.assertEqual(part.wipe, 'superblock')
+
+    def test_add_structure_no_fs(self):
+        disk = make_disk(self.fsc.model)
+        self.fsc._add_structure(
+            disk=disk,
+            offset=100,
+            size=2 << 20,
+            is_last=False,
+            structure=snapdapi.VolumeStructure(
+                type="83,0FC63DAF-8483-4772-8E79-3D69D8477DE4",
+                size=1 << 20,
+                filesystem=None))
+        [part] = disk.partitions()
+        self.assertEqual(part.size, 2 << 20)
+        self.assertEqual(part.fs(), None)
+        self.assertEqual(part.wipe, None)
+
+    def test_from_sample_data(self):
+        self.fsc.model = model = make_model(Bootloader.UEFI)
+        disk = make_disk(model)
+        with open('examples/snaps/v2-systems-unavailable.json') as fp:
+            self.fsc._system = snapdapi.snapd_serializer.deserialize(
+                snapdapi.SystemDetails, json.load(fp)['result'])
+        self.fsc.apply_system(disk.id)
+        self.assertEqual(
+            len(self.fsc._system.volumes['pc'].structure) - 1,
```

Comment:
The -1 is due to mbr and bios boot being split in the json, correct?
A quick comment would be nice.

Reply:
I've tried to make this clearer

Reply:
LGTM